power_grid_model_c/power_grid_model/include/power_grid_model/main_model.hpp

```diff
@@ -162,10 +162,18 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
         assert(construction_complete_);
 
         if constexpr (CacheType::value) {
+            constexpr auto comp_index = AllComponents::template index_of<CompType>();
+
+            auto& cached_sequence = std::get<comp_index>(cached_updated_sequence_indices_);
+            if (sequence_idx.empty()) {
+                main_core::get_component_sequence<CompType>(state_, begin, end, std::back_inserter(cached_sequence));
+            } else {
+                assert(cached_sequence.empty());
+                cached_sequence = sequence_idx;
+            }
+
             main_core::update_inverse<CompType>(
-                state_, begin, end,
-                std::back_inserter(std::get<AllComponents::template index_of<CompType>()>(cached_inverse_update_)),
-                sequence_idx);
+                state_, begin, end, std::back_inserter(std::get<comp_index>(cached_inverse_update_)), cached_sequence);
         }
 
         UpdateChange const changed = main_core::update_component<CompType>(state_, begin, end, sequence_idx);
```

Comment:
this line should also make use of the `cached_updated_sequence_indices_` if created